homeassistant/components/tuya/fan.py

```diff
@@ -0,0 +1,261 @@
+"""Support for Tuya Fan."""
+from __future__ import annotations
+
+import json
+import logging
+from typing import Any
+
+from tuya_iot import TuyaDevice, TuyaDeviceManager
+
+from homeassistant.components.fan import (
+    DIRECTION_FORWARD,
+    DIRECTION_REVERSE,
+    DOMAIN as DEVICE_DOMAIN,
+    SUPPORT_DIRECTION,
+    SUPPORT_OSCILLATE,
+    SUPPORT_PRESET_MODE,
+    SUPPORT_SET_SPEED,
+    FanEntity,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.util.percentage import (
+    ordered_list_item_to_percentage,
+    percentage_to_ordered_list_item,
+)
+
+from .base import TuyaHaEntity
+from .const import (
+    DOMAIN,
+    TUYA_DEVICE_MANAGER,
+    TUYA_DISCOVERY_NEW,
+    TUYA_HA_DEVICES,
+    TUYA_HA_TUYA_MAP,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+# Fan
+# https://developer.tuya.com/en/docs/iot/f?id=K9gf45vs7vkge
+DPCODE_SWITCH = "switch"
+DPCODE_FAN_SPEED = "fan_speed_percent"
+DPCODE_MODE = "mode"
+DPCODE_SWITCH_HORIZONTAL = "switch_horizontal"
+DPCODE_FAN_DIRECTION = "fan_direction"
+
+# Air Purifier
+# https://developer.tuya.com/en/docs/iot/s?id=K9gf48r41mn81
+DPCODE_AP_FAN_SPEED = "speed"
+DPCODE_AP_FAN_SPEED_ENUM = "fan_speed_enum"
+
+TUYA_SUPPORT_TYPE = {
+    "fs",  # Fan
+    "kj",  # Air Purifier
+}
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+):
+    """Set up tuya fan dynamically through tuya discovery."""
+    _LOGGER.debug("fan init")
+
+    hass.data[DOMAIN][entry.entry_id][TUYA_HA_TUYA_MAP][
+        DEVICE_DOMAIN
+    ] = TUYA_SUPPORT_TYPE
+
+    async def async_discover_device(dev_ids: list[str]) -> None:
```

Comment:
Please adjust here as in the light and climate PRs. All these platforms have the same pattern here so they should be the same, I think.
```suggestion
    @callback
    def async_discover_device(dev_ids: list[str]) -> None:
```